src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -299,50 +434,260 @@ EmbeddedAssemblies::gather_bundled_assemblies_from_apk (const char* apk, monodro
+	ssize_t nread = read (index_fd, data, data_size);
+	if (nread != static_cast<ssize_t>(data_size)) {
+		log_error (LOG_ASSEMBLY, "typemap: failed to read %u bytes from index file. %s", data_size, strerror (errno));
+		return nullptr;
+	}
+
+	uint8_t *p = data;
+	for (size_t i = 0; i < module_count; i++) {
+		memcpy (modules[i].module_uuid, p, UUID_SIZE);
+		modules[i].assembly_name = reinterpret_cast<char*>(p + UUID_SIZE);
+		p += entry_size;
+	}
+
+	return data;
+}
+
+uint8_t*
+EmbeddedAssemblies::typemap_load_index (int dir_fd, const char *dir_path, const char *index_path)
+{
+	log_debug (LOG_ASSEMBLY, "typemap: loading TypeMap index file '%s/%s'", dir_path, index_path);
+
+	TypeMapIndexHeader header;
+	size_t file_size;
+	int fd = -1;
+	uint8_t *data = nullptr;
+
+	if (!typemap_read_header (dir_fd, "TypeMap index", dir_path, index_path, MODULE_INDEX_MAGIC, header, file_size, fd)) {
+		goto cleanup;
+	}
+
+	module_count = header.entry_count;
+	modules = new TypeMapModule[module_count]();
+	data = typemap_load_index (header, file_size, fd);
+
+  cleanup:
+	if (fd >= 0)
+		close (fd);
+
+	return data;
+}
+
+bool
+EmbeddedAssemblies::typemap_load_file (BinaryTypeMapHeader &header, const char *dir_path, const char *file_path, int file_fd, TypeMapModule &module)
+{
+	size_t alloc_size = ADD_WITH_OVERFLOW_CHECK (size_t, header.assembly_name_length, 1);
+	module.assembly_name = new char[alloc_size];
+
+	ssize_t nread = read (file_fd, module.assembly_name, header.assembly_name_length);
+	module.assembly_name [header.assembly_name_length] = 0;
+	module.entry_count = header.entry_count;
+
+	log_debug (
+		LOG_ASSEMBLY,
+		"typemap: '%s/%s':: entry count == %u; duplicate entry count == %u; Java type name field width == %u; MVID == %s; assembly name length == %u; assembly name == %s",
+		dir_path, file_path, header.entry_count, header.duplicate_count, header.java_name_width,
+		mono_guid_to_string (header.module_uuid), header.assembly_name_length, module.assembly_name
+	);
+
+	alloc_size = MULTIPLY_WITH_OVERFLOW_CHECK (size_t, header.java_name_width + 4, header.entry_count);
+	module.java_name_width = header.java_name_width;
+	module.java_map = new uint8_t[alloc_size];
+	nread = read (file_fd, module.java_map, alloc_size);
+	if (nread != static_cast<ssize_t>(alloc_size)) {
+		log_error (LOG_ASSEMBLY, "typemap: failed to read %u bytes (java-to-managed) from module file %s/%s. %s", alloc_size, dir_path, file_path, strerror (errno));
+		return false;
+	}
+
+	module.map = new TypeMapModuleEntry[header.entry_count];
+	alloc_size = MULTIPLY_WITH_OVERFLOW_CHECK (size_t, sizeof(TypeMapModuleEntry), header.entry_count);
+	nread = read (file_fd, module.map, alloc_size);
+	if (nread != static_cast<ssize_t>(alloc_size)) {
+		log_error (LOG_ASSEMBLY, "typemap: failed to read %u bytes (managed-to-java) from module file %s/%s. %s", alloc_size, dir_path, file_path, strerror (errno));
+		return false;
+	}
+
+	// alloc_size = module.java_name_width + 1;
```

Comment:
Why have this commented-out code?

Reply:
I use it for debugging, don't want to remove it for now